#2366 Make RPE date range configurable

shaunxp20 opened the pull request:
This will make the RPE date range dynamic via new configuration settings:
```
DATES_REGIONAL_PITCH_EVENTS_BEGINS_YEAR
DATES_REGIONAL_PITCH_EVENTS_BEGINS_MONTH
DATES_REGIONAL_PITCH_EVENTS_BEGINS_DAY

DATES_REGIONAL_PITCH_EVENTS_ENDS_YEAR
DATES_REGIONAL_PITCH_EVENTS_ENDS_MONTH
DATES_REGIONAL_PITCH_EVENTS_ENDS_DAY
```

shaunxp20 commented on spec/technovation/configuration_settings_spec.rb:
```diff
@@ -0,0 +1,51 @@
+  require "rails_helper"
```
Let me know what you think about these specs, @stenington 🙏 

My thought was to ensure these configuration settings are setup correctly before deploying to staging/prod, and if they're not then the the tests will fail, which will break the build which will stop the deploy. 🤔 🤷‍♂ I also added a few simple specs to validate the date formats too.

If we like this, maybe we can add another issue for the rest of them?

A couple of random thoughts:

I struggled w/ referring to them as configuration settings vs environment variables, but I think it makes as configuration settings.

I like having specs explicit/wordy, and that's what I started doing, but your specs for the `curriculum-checks` specs inspired me to make them more dynamic. I think it's not so bad in this case. 🤔 🤷 😄

shaunxp20 replied:
Hmm, it looks like the build is failing b/c the environment variables need to be setup in CircleCI as well. 🤔 😒 I'll wait to see if we want to go down this path before adding them/moving forward. 👍 😃

stenington replied:
Well, the problem here is that we don't run tests in production, so these tests won't actually ensure that our production environment is correct, just that our test environment is correct (whether that's a local dev environment or CI). While that may be a little bit useful, I worry it gives a false sense of confidence that things will work in production without thinking about it carefully. What do you think?

shaunxp20 replied:
> Well, the problem here is that we don't run tests in production, so these tests won't actually ensure that our production environment is correct

Ah! Good to know! I guess I went a little off reservation!! 😂 😊 😁 FWIW, I would be interested in exploring the pros/cons of running some of our smoke tests in production. I think running these (and some others) would provide value. 😄 

And as you pointed out, these new specs would ensure our local environments are setup correctly (which I think would add some value), but falls short of my main goal of ensuring staging/prod is setup correctly (where it stand now anyway). I think I'll remove/stash these specs for now, to not give the false sense of confidence (I agree), and it would be consistent w/ our other ENVs in that none of the others are currently being tested. 😀 😁 

Anywho, I found something else that we might be able to use w/ our `dotenv-rails` gem. There's a section on [Required Keys](https://github.com/bkeepers/dotenv#required-keys):

> If a particular configuration value is required but not set, it's appropriate to raise an error.
```
Dotenv.require_keys('BLAH')
```
> If any of the configuration keys above are not set, your application will raise an error during initialization. This method is preferred because it prevents runtime errors in a production application due to improper configuration.

Maybe this could meet the "fail fast" requirement too. 🤔 Thoughts?

stenington replied:
Yeah, that's a possibility. Right now I believe we don't even install `dotenv-rails` in staging or production, but I'm very open to something like an env initializer that blows up on missing required settings regardless of environment, that would be a great thing to have.

stenington replied:
We discussed this in stand briefly, but writing it here as well: It occurred to me that if we were able to run ENV checks as part of a Heroku build, we might be able to prevent a deployment from evening happening if certain criteria weren't satisfied, which is sort of nicer than deploying an app that then immediately refuses to run. Something to look into maybe... I'm not entirely sure how it might be done or if there's a downside I'm not thinking of.

shaunxp20 replied:
I'm not sure if this is what you were thinking, but I'll throw this idea out there... Apparently the ENV vars on Heroku are available via an API: [Config Vars Info for App](https://devcenter.heroku.com/articles/platform-api-reference#config-vars)

We could pull down the ENV vars via the API and do some validations against them, then make this a part of the build process (so the build will fail if the validations don't pass). I have no idea if this is what you had in mind! 😊 😄 🤔 🏖 🍹 🌞

shaunxp20 replied:
@stenington and I discussed this. In summary, I'm going to open a new issue to test our ENV vars; it's a little out of scope for this PR, and that way it doesn't hold up this PR either. 👍 ▶️

shaunxp20 commented on app/javascript/ra/events/EventForm.vue:
```diff
@@ -462,6 +462,18 @@
+      minDate() {
+        return process.env.DATES_REGIONAL_PITCH_EVENTS_BEGINS_YEAR + '-' +
+          process.env.DATES_REGIONAL_PITCH_EVENTS_BEGINS_MONTH + '-' +
+          process.env.DATES_REGIONAL_PITCH_EVENTS_BEGINS_DAY
```
I tried string interpolation here, but (imho) it didn't make it any more readable. 😕 Open to suggestions!

shaunxp20 replied:
Having one environment variable contain the entire date would probably help, but I was just following the convention of our other environment variables where the date is split up. 💇‍♂ 💅 💇 🤷‍♂ 😺

stenington replied:
I had absolutely no idea you could use `process.env` from front end code, so this is interesting. 

I was assuming we'd read the config [somewhere in here](https://github.com/Iridescent-CM/technovation-app/blob/qa/app/constants/important_dates.rb) and render it as a data attribute onto a page or whatever it is this Vue app otherwise does to get initialized. (Note, I'd prefer that Vue apps initialize with an initial ajax call but many of them don't currently.)

This is in a way much cleaner, but at the cost of decentralizing some of our reading-dates-from-the-env setup. I guess my question is whether or not this will "fail fast" if that config is missing? Or will the front end just display a broken date picker? What if instead of string concatenation you pass the variables to `new Date(...)`, would that fail fast if the config wasn't there, or will the app still run with a broken date picker?

shaunxp20 commented:
@stenington Unfortunately, no tests for this PR. 😭 😞 But I did verify the functionality again locally. I also added the appropriate config/ENV key/values to both staging and production, and added QA steps in [#2366](https://github.com/Iridescent-CM/technovation-app/issues/2366#issuecomment-587923099). Let me know what you think! 😄

stenington commented:
👍 Works for me!
